source/common/filter/tcp_proxy.cc

```diff
@@ -306,26 +328,35 @@ void TcpProxy::onUpstreamData(Buffer::Instance& data) {
 }
 
 void TcpProxy::onUpstreamEvent(Network::ConnectionEvent event) {
+  bool connecting = false;
+
+  // The timer must be cleared before, not after, processing the event because
+  // if initializeUpstreamConnection() is called it will reset the timer, so
+  // clearing after that call will leave the timer unset.
```

Comment:
Can you add an `ASSERT(connect_timeout_timer_ == nullptr)` in `initializeUpstreamConnection` before you assign it there?

Reply:
Done